codegenerator/cli/templates/static/codegen/src/EventFetching.res

```diff
@@ -72,91 +72,79 @@ let makeCombinedEventFilterQuery = (
-      let queryTimoutPromise =
-        Time.resolvePromiseAfterDelay(~delayMilliseconds=sc.queryTimeoutMillis)->Promise.then(() =>
-          Promise.reject(
-            QueryTimout(
-              `Query took longer than ${Belt.Int.toString(sc.queryTimeoutMillis / 1000)} seconds`,
-            ),
-          )
+): promise<eventBatchQuery> => {
+  let rec executeQuery = (~suggestedBlockInterval): promise<eventBatchQuery> => {
+    //If the query hangs for longer than this, reject this promise to reduce the block interval
+    let queryTimoutPromise =
+      Time.resolvePromiseAfterDelay(~delayMilliseconds=sc.queryTimeoutMillis)->Promise.then(() =>
+        Promise.reject(
+          QueryTimout(
+            `Query took longer than ${Belt.Int.toString(sc.queryTimeoutMillis / 1000)} seconds`,
+          ),
         )
+      )
 
-      let upperBoundToBlock = fromBlockRef.contents + blockInterval - 1
-      let nextToBlock =
-        Pervasives.min(upperBoundToBlock, toBlock)->Pervasives.max(fromBlockRef.contents) //Defensively ensure we never query a target block below fromBlock
-      let logsPromise =
-        makeCombinedEventFilterQuery(
-          ~contractInterfaceManager,
-          ~fromBlock=fromBlockRef.contents,
-          ~toBlock=nextToBlock,
-          ~provider,
-          ~logger,
-        )->Promise.thenResolve(logs => (logs, nextToBlock - fromBlockRef.contents + 1))
-
-      [queryTimoutPromise, logsPromise]
-      ->Promise.race
-      ->Promise.catch(err => {
-        logger->Logging.childWarn({
-          "msg": "Error getting events, will retry after backoff time",
-          "backOffMilliseconds": sc.backoffMillis,
-          "err": err,
-        })
+    let suggestedToBlock = fromBlock + suggestedBlockInterval - 1
+    let queryToBlock = Pervasives.min(suggestedToBlock, toBlock)->Pervasives.max(fromBlock) //Defensively ensure we never query a target block below fromBlock
+    let latestFetchedBlockPromise = loadBlock(queryToBlock)
+    let logsPromise = makeCombinedEventFilterQuery(
+      ~contractInterfaceManager,
+      ~fromBlock,
+      ~toBlock=queryToBlock,
+      ~provider,
+      ~logger,
+    )->Promise.then(async logs => {
+      let executedBlockInterval = queryToBlock - fromBlock + 1
+      // Increase the suggested block interval only when it was actually applied
+      // and we didn't query to a hard toBlock
+      let nextSuggestedBlockInterval = if executedBlockInterval >= suggestedBlockInterval {
+        // Increase batch size going forward, but do not increase past a configured maximum
+        // See: https://en.wikipedia.org/wiki/Additive_increase/multiplicative_decrease
+        Pervasives.min(executedBlockInterval + sc.accelerationAdditive, sc.intervalCeiling)
+      } else {
+        suggestedBlockInterval
+      }
+      {
+        logs,
+        nextSuggestedBlockInterval,
+        latestFetchedBlock: await latestFetchedBlockPromise,
```

Comment:
Why is this latestFetchedBlock moved here? Any reason to not keep the first and last block of the range being queried together in the code? Since those are both needed for the reorg detection and I don't see an additional use for it here?

Reply:
This is because now we don't accumulate responses until the target block, and we don't know beforehand what will be the actual to block. I moved it here to keep the pervious logic of calling for the block data before doing the get_logs request, I think it's more reliable in terms of the reorg detection. But I can move the block data call after we've received the get_logs response, this will be even cheaper in terms of rpc calls.

Reply:
Ah yes I see! I think it doesn't matter either way I just didn't immediately see why it was moved 👍🏼